sdk/python/feast/pyspark/launchers/aws/emr.py

```diff
@@ -0,0 +1,244 @@
+        """
+        Submit EMR job using a new or existing cluster.
+
+        Returns a tuple of JobId, StepId
+        """
+
+        emr = self._emr_client()
+
+        if self._existing_cluster_id:
+            step["ActionOnFailure"] = "CONTINUE"
+            step_ids = emr.add_job_flow_steps(
+                JobFlowId=self._existing_cluster_id, Steps=[step],
+            )
+            return EmrJobRef(self._existing_cluster_id, step_ids["StepIds"][0])
+        else:
+            assert self._new_cluster_template is not None
+            jobTemplate = self._new_cluster_template
+            step["ActionOnFailure"] = "TERMINATE_CLUSTER"
+
+            jobTemplate["Steps"] = [step]
+
+            if self._emr_log_location:
+                jobTemplate["LogUri"] = os.path.join(
+                    self._emr_log_location, _random_string(5)
+                )
+
+            job = emr.run_job_flow(**jobTemplate)
+            return EmrJobRef(job["JobFlowId"], None)
+
+    def historical_feature_retrieval(
+        self, job_params: RetrievalJobParameters
+    ) -> RetrievalJob:
+
+        with open(job_params.get_main_file_path()) as f:
+            pyspark_script = f.read()
+
+        pyspark_script_path = _s3_upload(
+            BytesIO(pyspark_script.encode("utf8")),
+            local_path="historical_retrieval.py",
+            remote_path_prefix=self._staging_location,
+            remote_path_suffix=".py",
+        )
+
+        step = _historical_retrieval_step(
+            pyspark_script_path, args=job_params.get_arguments()
+        )
+
+        job_ref = self._submit_emr_job(step)
+
+        return EmrRetrievalJob(
+            self._emr_client(),
+            job_ref,
+            os.path.join(job_params.get_destination_path(), _random_string(8)),
+        )
+
+    def offline_to_online_ingestion(
+        self, ingestion_job_params: IngestionJobParameters
+    ) -> IngestionJob:
+        """
+        Submits a batch ingestion job to a Spark cluster.
+
+        Raises:
+            SparkJobFailure: The spark job submission failed, encountered error
+                during execution, or timeout.
+
+        Returns:
+            IngestionJob: wrapper around remote job that can be used to check when job completed.
+        """
+
+        jar_s3_path = _upload_jar(
+            self._staging_location, ingestion_job_params.get_main_file_path()
+        )
+        step = _sync_offline_to_online_step(
+            jar_s3_path,
+            ingestion_job_params.get_feature_table_name(),
+            args=ingestion_job_params.get_arguments(),
+        )
+
+        job_ref = self._submit_emr_job(step)
+
+        return EmrIngestionJob(self._emr_client(), job_ref)
+
+    def stage_dataframe(
+        self, df: pandas.DataFrame, event_timestamp: str, created_timestamp_column: str
+    ) -> FileSource:
+        with tempfile.NamedTemporaryFile() as f:
+            df.to_parquet(f)
```

Comment:
@pyalex As of Feast 0.7 the dataframe is typically converted to pyarrow / Avro format. I am wondering, if we should make parquet the standard staging format as oppose to pyarrow / Avro?  Since Spark doesn't support Avro format out of the box.

Reply:
> As of Feast 0.7 the dataframe is typically converted to pyarrow / Avro format

In which instances do you see that? Uploading to BQ?

Reply:
https://github.com/feast-dev/feast/blob/v0.7.1/sdk/python/feast/loaders/file.py#L29

Reply:
we already write parquet in `client.ingest` but with DataFormat PR we'll have to support all formats that user may specify as Batch Source format, since `client.ingest` writes to existing batch source

Reply:
@pyalex  In the case of historical feature retrieval, where the user input is Panda dataframe, do we actually want to make the user specify the format of the entity source file? Or simply standardized to parquet?

Reply:
Didn't we previously agree to only support parquet?